crates/sparrow-backend/src/pipeline_schedule.rs

```diff
@@ -0,0 +1,163 @@
+use sparrow_core::debug_println;
+use sparrow_physical::{Pipeline, Step, StepKind};
+
+const DEBUG_SCHEDULING: bool = false;
+
+/// Determine the pipeline each step should be part of.
+///
+/// This is computed in a bottom-up manner:
+///
+/// 1. Each leaf (input) is the start of a separate pipeline.
+/// 2. Certain step kinds are "pipeline breaking", which means
+///    they start a new pipeline. For instance, `with_key`.
+/// 3. Any other operation with a single input is part of the
+///    the same pipeline as the input.
+/// 4. Any other operation is a separate pipeline.
+pub fn pipeline_schedule(steps: &[Step]) -> Vec<Pipeline> {
+    // Compute how many times each step is referenced.
+    // Any step referenced multiple times must end a pipeline.
+    let mut references = vec![0; steps.len()];
+    for step in steps {
+        for input in &step.inputs {
+            references[input.0] += 1;
+        }
+    }
+
+    // Then compute the assignments for each step.
+    let mut assignments = Vec::with_capacity(steps.len());
+
+    let mut pipelines = Vec::new();
+    for (index, step) in steps.iter().enumerate() {
+        let break_pipeline = if step.inputs.len() != 1 {
+            debug_println!(
+                DEBUG_SCHEDULING,
+                "Step {index} is new pipeline since it has multiple inputs"
+            );
+            true
+        } else if references[step.inputs[0].0] > 1 {
+            debug_println!(
+                DEBUG_SCHEDULING,
+                "Step {index} is new pipeline since it's only input ({}) is referenced {} times",
+                step.inputs[0].0,
```

Comment:
The use of `.0` (to get the `usize` out of the newtype wrappers) and `.into()` to convert `usize` to the newtype wrappers) is kind of a pain. I'm partway inclined to say we should just use `usize` in the physical plan for simplicity. Thoughts?

Reply:
I thought there was a way to get the wrapper to be "transparent" with some `repr` trick. Does that not work in this case?

Reply:
So, `repr(transparent)` just means that in assembly, there is no difference between the struct and the field -- so it's a "zero cost" wrapper, at runtime.

But, the whole value of the type wrapper is that it requires you to explicitly convert in/out. So we *could* add a `From<ExprId> for usize`, which would let both conversions be `x.into()`, but there will always need to be an explicit conversion.

The alternative would be a type alias `type ExprId = usize`, but that is basically "erased" at compile time, I believe, so it provides no safety benefits.

Reply:
Ah got it, thanks

Reply:
So I guess, we have the following options:

1. Keep the wrappers. We'll always need to do `2.into()` and `some_array[2.into()]`, unless we create our own "wrapper aware vector" so we can override the indexing to do the conversion. Seems to be heavyweight, but doable. For instance `ExprId` and `ExprIdVec`, etc.
2. Drop the wrappers. Use `usize`. Use good naming and trust ourselves.
3. Use type aliases. Keeps a unique name, but doesn't provide any protections.

I'm generally torn between 1 and 2 -- I don't think 3 buys us much. Thoughts?